### ISSUE TITLE
v3 POST notification - implement basic send logic

### ISSUE THREAD

**Issue by kalbfled** [department-of-veterans-affairs/notification-api#1361]
## User Story - Business Need

This is a continuation of #1360 to implement a subset of the functionality necessary to send SMS and e-mail messages.  The previous ticket left a [TODO](https://github.com/department-of-veterans-affairs/notification-api/blob/master/app/v3/notifications/rest.py#L99) to pass validated request data to Celery for additional processing as described in the "Celery tasks" portion of the [v3 flowchart](https://github.com/department-of-veterans-affairs/vanotify-team/blob/master/Engineering/v3_post_notifications_flow.md).

The big picture goal for this ticket is that sending SMS and e-mail notifications elegantly works for one provider (per notification type).  This should be enough to reveal the general structure of the new v3 implementation, and that will provide the basis for additional tickets.

- [x] Ticket is understood, and QA has been contacted (if the ticket has a QA label).

### User Story(ies)

**As a** Notify developer,
**I want** creating and sending notifications to be fast
**So that** we can scale to many more messages sent.

**As a** Notify client,
**I want** useful notification status reasons when send a notification fails
**So that** I have the ability to troubleshoot.

### Additional Info and Resources

- [CeleryQ documentation](https://docs.celeryq.dev) (ensure you reference the version actually in use)
- [ORM Querying Guide](https://docs.sqlalchemy.org/en/20/orm/queryguide/index.html)

**Calls to third parties (ex. VA Profile ID look up, deliver notification to provider, etc.) should be standalone Celery tasks.**  We don't want them to block or result in duplicate computation when they need to retry.  For example, don't check service permissions in the same task as a request to VA Profile.  If these are one task and the request times-out, the permission checks, which already succeeded, will be repeated.

**Celery tasks should not default to retrying.**  Only failures outside of our control (i.e. third party time-outs) should result in a retry.

## Engineering Checklist

- [x] Modify the signature of the function [v3_send_notification](https://github.com/department-of-veterans-affairs/notification-api/blob/master/app/v3/notifications/rest.py#L61) to take a second parameter "service_data" of type [ServiceData](https://github.com/department-of-veterans-affairs/notification-api/blob/master/app/service/service_data.py#L108).
- [x] Above the function modified above, modify the two route handlers that call the function so they pass app.authenticated_service to the function.  (Example of usage in v2 [here](https://github.com/department-of-veterans-affairs/notification-api/blob/master/app/v2/notifications/post_notifications.py#L122).)
- [x] In the new file app/celery/v3/notification_tasks.py, create a Celery task like this: `def v3_process_notification(request_data: dict, service_data: ServiceData)`.
  - Open question: Should [NotifyCelery](https://github.com/department-of-veterans-affairs/notification-api/blob/master/app/celery/celery.py#L49) continue to be used?  It has side-effects.
- [x] At the [TODO location](https://github.com/department-of-veterans-affairs/notification-api/blob/master/app/v3/notifications/rest.py#L99) in v3_send_notification, initiate a celery task like this: `v3_process_notification.delay(request_data, service_data)`.

**With the above tasks complete, this ticket is less prescriptive.  Given the "big picture goal" described above, and **without using code in app/clients/** to the maximum extent practical:**

- [x] Implement sending SMS with 1 provider (pinpoint)
- [x] Implement sending e-mail with 1 provider (AWS ses)

**At a minimum, new unit tests should verify the following:**

- [x] Send SMS happy path with pinpoint (includes appropriate status updates)
- [x] Send e-mail happy path with SES (includes appropriate status updates)
- [x] Service permissions work as with v2 (ex. service has permission to send an e-mail notification)
- [x] The template_id in the request body must correspond to a template with the same notification type specified in the request (i.e. template.template_type == request_body.notification_type).  Failure should result in a new Notification instance with permanent failure status.
- [ ] For an SMS notification with request data that does not include an sms_sender_id, use the template's default sms_sender_id.  If the latter isn't available, the notification is invalid.
- [x] All permanent failures result in a new Notification instance in the database with an appropriate reason for the failure.
- [ ] A notification should not be valid if it specifies a template_id or sms_sender_id not owned by the authenticated service
- [ ] Document all error types, statuses, and status reasons in a github file and have @k-macmillan and @mjones-oddball sign off on status reason wording

**Create new tickets as necessary for completing the implementation:**

- [ ] We want to return an http response without additional database queries even for a notification that specifies a nonexistent template ID, but persisting the notification requires a valid template ID.  Either make the Notification.template_id column nullable,  create a default template, etc.  **Handle this before any other follow-on work.**  See #1493.
- [ ] The implementation to send e-mail contains a lot of commented-out code, which means the behavior does not match v2.  Fix this.  (The commented-out code doesn't work as written.)
- [ ] Implement personalization.  As currently written, the recipient will receive the notification's unaltered template content, including placeholder variables.
- [ ] Sending with scheduled_for (see [here](https://docs.celeryq.dev/en/stable/userguide/calling.html#eta-and-countdown))
- [ ] Sending with additional providers . . .
- [ ] Statsd reporting works for v3 (It might be okay for this to use existing code.)
- [ ] Onsite notifications occur when appropriate.  See [vanotify-team 975](https://github.com/department-of-veterans-affairs/vanotify-team/issues/975).
- [ ] A request that includes a recipient ID and direct contact info (phone number or e-mail address) does not use the recipient ID to lookup direct contact info (which is what v2 does).
- [ ] Others . . . (add them)

## Acceptance Criteria
- [ ] Follow the [logging document](https://github.com/department-of-veterans-affairs/vanotify-team/blob/master/Engineering/logging_policy.md)
- [ ] All the unit tests above pass
- [ ] Postman send SMS message --> message received
- [ ] Postman send e-mail message --> message received
- [ ] All queries utilize the new (SQL Alchemy 2.0) [query syntax](https://flask-sqlalchemy.palletsprojects.com/en/3.1.x/quickstart/#query-the-data).
- [ ] New code does not use any dao, v1, or v2 functions (make all database queries directly)
- [ ] New code uses the database read-only instance when appropriate
- [ ] New code uses the Python "requests" package directly rather than using existing "client" code (which does not consistently handle errors)

### Requirements

- New v3 email `/v3/notifications/email` and sms `/v3/notifications/sms` endpoints will accept requests with the listed functional fields below and will support notification delivery with a defined provider (hardcoded SES and pinpoint respectively), logging (can be verified in cloudwatch and/or datadog), and associated status updates in DB only (can be verified by status endpoint).

    - Email Fields:
        - Functional
            - template_id
            - billing_code
            - email_address
        - Non Functional
            - reference
            - client_reference
            - recipient_identifier
                - id_type
                - Id_value
            - scheduled_for
            - email_reply_to_id
            - personalisation

    - SMS Fields:
        - Functional
            - template_id
            - billing_code
            - phone_number
            - sms_sender_id
                - Default logic
                - Sender id logic for pinpoint only
                - Twilio logic will be deferred to when we support other providers
                - Error handling required here!
                    - We’ll be using permanent_failure status and giving realistic reasons. 
        - Non Functional
            - reference
            - client_reference
            - recipient_identifier
                - id_type
                - Id_value
            - scheduled_for
            - personalisation
    - Notification Statuses: Will be updated in DB and trigger callbacks if enabled
        - Success: When request is accepted by the provider status is Sending
            - Please note: V3 does not change existing provider code/logic so after the request goes to the provider, subsequent processing and status updates are unchanged
        - Errors:
            - API request body validation error handling and associated status updates
                - Status = permanent_failure
        - Third-party connection, rate limiting, and timeout errors - all retryable according to existing retry logic/code
            - QA will only test SES and pinpoint at this point. VAPRO, MPI, and Twilio will be handled in later tickets.
        - Delivery Status Callbacks will be included for all errors accounted for in this part of the work

- **NOT FUNCTIONAL YET**
    - Logic for non-functional fields
    - Onsite notification logic
    - Additional provider logic (e.g. Twilio)
    - Sms sender logic for Twilio

## QA Considerations
- [ ] determine any changes to the QA Suite based on convo with dev
- [ ] regression passes

Note from Product: QA might want to test callbacks until the end of all the logic work instead of validating in this ticket. Your call!

## Out of Scope

- Do not modify POST notifications code for v1 or v2.
- Do not implement anything for which a new ticket is to be created (as listed above)
- Do not implement logic for additional providers; only 1 SMS provider and 1 e-mail provider

**Comment by kalbfled** [comment]
I'm thinking that having a bunch of different tasks might be undesirable.  It might be better to have 1 task that makes async calls to 3rd parties.

**Comment by kalbfled** [comment]
Note to self: For a SMS notification, look for a default ServiceSmsSender if sms_sender_id wasn't provided in the request data.

**Comment by kalbfled** [comment]
@cris-oddball I haven't forgotten about this.  I will update it Monday morning.

**Comment by kalbfled** [comment]
Team,

I will be absent during refinement for this ticket.  My intent is that I work on this ticket for an entire sprint (up to the QA cutoff), which makes this an 8 by our definition.  This is a time box.  Whatever I can get done in one sprint is what this ticket is.  If I don't think I will be able to complete a feature in time, I will create a follow-up ticket.

I feel that this approach is appropriate for this task because it is mostly new code.  Once a scaffold is in place, it will be easier to estimate completion times for specific features.  For example, "Implement scheduled sending."

**Comment by k-macmillan** [comment]
Bear in mind that phone/email are used for whitelist (restricted). Recipient ID is not eligible for restricted sends. Simulated is not going to be carried forward with v3.

**Comment by mjones-oddball** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/VA-Notify-620d21369d810a00146ed9c8/issues/department-of-veterans-affairs/notification-api/1361?planning-poker) with Zenhub @babyjunior56 @EvanParish @k-macmillan @kalbfled @ldraney @nikolai-efimov

**Comment by kalbfled** [comment]
I discovered that I do not need to make any additional database queries to verify send permissions.  Therefore, contrary to the v3 flowchart, I verify the permissions before returning an HTTP response rather than in a Celery task.  This does not add significant overhead to the route handlers, and it has the benefit of of giving users immediate feedback.

I also implemented a 501 ("Not Implemented") error handler.  That is the expected response for a request with "scheduled_for", which we plan to implement later.

**Comment by kalbfled** [comment]
@cris-oddball I pushed branch "1361-v3-part2" yesterday.  I have not yet implemented the send logic, but some things are testable now:
- [ ] Service without e-mail permission trying to send e-mail gets a 403
- [ ] Service without SMS permission trying to send SMS gets a 403
- [ ] Request with "scheduled_for" gets a 501 response
- [ ] Request with template ID for nonexistent template gets a 202 but fails downstream with an error log message (I discussed this with @k-macmillan and @mjones-oddball.  This is a priority for follow-up work.  See the "new tickets" section of this ticket's description.)
- [ ] Request with a template ID not owned by the service gets a 202 and creates a new Notification with "permanent failure" status and associated reason.  (I have an xfail'ed unit test for this; will fix later.  I expect the functionality to work now.)
- [ ] Request with a template ID of a different type (i.e. e-mail) than the notification type (i.e. SMS) gets a 202 and creates a new Notification with "permanent failure" status and associated reason

Also not implemented:
- Testing permission for international SMS

**Comment by k-macmillan** [comment]
task.delay is a wrapper for task.apply_async (`delay` doesn't allow execution options). All of our apply_async calls involve specifying a queue. I went looking for our default queue and I could not find one so I changed `delay` to `apply_async` to validate the assumption. We're now progressing through the celery tasks.
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBLzBZQVE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--45ec5b67026ea2273e151ade848a228cd09ca109/image.png)

Looked into this further and it appears we should be specifying [task_routes](https://docs.celeryq.dev/en/stable/userguide/configuration.html#task-routes) to identify which queue things should go to. There is a `task_queues` parameter in the configuration section of our code, but the celery docs advise not using it, and instead using task_routes.

**Comment by k-macmillan** [comment]
Tried to get the minimum steps necessary to actually send an email without calling all the existing methods. Met a lot of resistance in the form of getting the right data out of the returned query.

**Comment by kalbfled** [comment]
@cris-oddball I have deployed the code to Dev successfully and gotten a 202 response.  POSTing a notification and then GETting the notification reveals a "technical failure" status with the "reason" that sending isn't implemented yet.  I'm working on the actual send logic today, beginning with SMS.  The checkbox list I created in a previous comment should now be applicable.

I still need to implement more robust error handling.  Right now, for example, POSTing with a syntactically valid but nonexistent sms_sender_id will cause a traceback to show up in the logs.  In general, nonexistent foreign keys will cause database integrity errors.

**Comment by cris-oddball** [comment]
@kalbfled For the [last test in the comment above](https://app.zenhub.com/workspaces/va-notify-620d21369d810a00146ed9c8/issues/gh/department-of-veterans-affairs/notification-api/1361#issuecomment-1748888946), it fails for sending an email with an sms template ID

When sending an email with an sms template, GET notifications returns a 500 Internal Server Error. It works as expected for sending an sms with an email template.

- Request with a template ID of a different type (i.e. e-mail) than the notification type (i.e. SMS) gets a 202 and creates a new Notification with "permanent failure" status and associated reason

  - v3 POST email with sms template id b96b27e3-9553-4b7d-8b1b-dcecef3ca6d9
  	- "id": "d8f99296-c8bc-4b7a-afd1-045fb55f1246" 202
  	- GET notification 500 Internal Server Error
```
{
    "message": "Internal server error",
    "result": "error"
}
```

  - v3 POST sms with email template id 18e95471-0488-4ce1-ac80-33701fe7fe0a
  	- "id": "dd1cca4e-af43-4825-921b-97cf8742c10a" 202
```
    "status": "permanent-failure",
    "status_reason": "The template type does not match the notification type.",
```

**Comment by kalbfled** [comment]
I am able to recreate the above error locally.  This is the full traceback:
```
ERROR 2023-10-11 18:45:17,297 /app/app/__init__.py:443: 'SMSMessageTemplate' object has no attribute 'subject'
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/flask/app.py", line 867, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python3.8/site-packages/ddtrace/contrib/flask/wrappers.py", line 88, in wrapper
    return func(pin, wrapped, instance, args, kwargs)
  File "/usr/local/lib/python3.8/site-packages/ddtrace/contrib/flask/patch.py", line 512, in _patched_request
    return wrapped(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/flask/app.py", line 852, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
  File "/usr/local/lib/python3.8/site-packages/ddtrace/contrib/flask/wrappers.py", line 61, in patch_func
    return _wrap_call(
  File "/usr/local/lib/python3.8/site-packages/ddtrace/contrib/flask/wrappers.py", line 52, in _wrap_call
    return wrapped(*args, **kwargs)
  File "/app/app/v2/notifications/get_notifications.py", line 16, in get_notification_by_id
    return jsonify(notification.serialize()), 200
  File "/app/app/models.py", line 1582, in serialize
    "subject": self.subject,
  File "/app/app/models.py", line 1476, in subject
    return template_object.subject
AttributeError: 'SMSMessageTemplate' object has no attribute 'subject'
```

**Comment by cris-oddball** [comment]
[SPIKE](https://app.zenhub.com/workspaces/va-notify-620d21369d810a00146ed9c8/issues/gh/department-of-veterans-affairs/notification-api/1493) to handle above issue.

**Comment by kalbfled** [comment]
Using the Dev environment, I successfully sent SMS notification c8d927f3-ca5f-4cf0-8855-611b96c027ab using the v3 route, but personalization doesn't work.  The markup and template variables were in the message I received.  Personalization is out of scope for this ticket, but I hoped it would just work.  I also noted that the notification's "status" remains "sent" even though I received it.  I expected status updates to work without additional effort.

**Comment by kalbfled** [comment]
Using the Dev environment, I successfully sent e-mail notification 75146868-d8ec-4601-997b-ea93fa1a2963 using the v3 route.  The implementation is minimal and needs additional work to restore all the features of v2.  I will create a new ticket for that.

**Comment by cris-oddball** [comment]
**TESTING NOTES ON [PR #1](https://github.com/department-of-veterans-affairs/notification-api/pull/1483)**

Noting that this is a WIP and that the code is well commented with TODOs for the remaining PRs/tickets to complete v3, here are some happy path notes:

1. send v3 email
payload
```
{
    "template_id": "{{email-template-id}}",
    "email_address": "<redacted>"
}
```
Received 202 Accepted
```
{
    "id": "d2d9f48c-d1aa-427b-948f-679cd10821f7"
}
```
Received the email
GET v2/notifications/d2d9f48c-d1aa-427b-948f-679cd10821f7
```
    "status": "sent",
    "status_reason": null,
```
This is a v3 bug that Dave will continue to work on. Logs:
```
2023-10-12T20:41:10 delivery delivery INFO no-request-id "Duplicate callback received. Notification id d2d9f48c-d1aa-427b-948f-679cd10821f7 received a status update to delivered 0:00:08.103917 after being set to sent. email sent by ses" [in /app/app/dao/notifications_dao.py:794]
[2023-10-12 20:41:10,244: INFO/ForkPoolWorker-4] Duplicate callback received. Notification id d2d9f48c-d1aa-427b-948f-679cd10821f7 received a status update to delivered 0:00:08.103917 after being set to sent. email sent by ses
```

2. send v3 email
Must be sent with sms_sender_id and that must be the default on the service.
payload:
```
{
    "template_id": "{{sms-template-id}}",
    "phone_number": "<redacted>",
    "sms_sender_id": "{{sms-sender-id}}"
}
```
Received 202 accepted
Received text
```
    "status": "sent",
```

If attempting to send without the sms_sender_id, the GET notification response is
```
    "status": "technical-failure",
    "status_reason": "Default logic for sms_sender_id is not yet implemented.",
```

If attempting to send with the alternative pinpoint sms_sender_id on the service, the GET notification response is
```
    "status": "created",
    "status_reason": null,
```
This is also a v3 bug.

Approving PR #1 as close enough for R&R (3 chords and the truth, which is really Blue, but hey).
Will not include this code in the release.

**Comment by cris-oddball** [comment]
Dave hopes to address Kyle's concerns before Planning

**Comment by kalbfled** [comment]
Rather than create a bunch of new tickets now--with content subject to change as other development completes--I created [v3 TODO](https://github.com/department-of-veterans-affairs/notification-api/wiki/v3-TODO) in the repo's Wiki.
